### ISSUE TITLE
WordPress Patient Portals are really really broken

### ISSUE THREAD

**Issue by tywrenn** [openemr/openemr#3571]
### Describe the bug

All WordPress up-for-grabs demos, including 5.0.2, are broken. A 500 error occurs when trying to login to the admin area.

tbh WordPress has been broken for awhile and we need to address it now. I just tried installing the WordPress patient portal and it's not at all updated.

@sunsetsystems seems to have done some work on this -- any help is appreciated

**Comment by tywrenn** [comment]
@bradymiller This is a sad state of affairs -- this hasn't been updated since 2014 ;-;

**Comment by sunsetsystems** [comment]
I think we can remove those demos. They get very little interest.

**Comment by tywrenn** [comment]
@sunsetsystems https://community.open-emr.org/t/wordpress-as-portal-patient/12992

**Comment by tywrenn** [comment]
Also linking this to current convo that I just saw:
https://community.open-emr.org/t/how-to-connect-wordpress-with-openemr/8937/30

**Comment by gutiersa** [comment]
Ninja forms is pretty broken

**Comment by bradymiller** [comment]
Since the onsite portal is so nice now, would it make sense to consider:
1. removing wordpress portal
2. removing offsite portal (@arnabnaha , are you still using this portal?)
(since neither of these have been maintained for several years)

**Comment by sjpadgett** [comment]
I actually been meaning to emulate some features from WP portal that are missing from onsite.

**Comment by arnabnaha** [comment]
Hi Brady,

I am only using the onsite portal. Not the wordpress or offsite...can
easily remove it after jerry introduces some of the wp pprtal features to
onsite.

**Comment by sjpadgett** [comment]
It's been 5 years since I last looked at the WP Portal. Could anyone that has some experience with Onsite and WP give me some idea what features would need to be replaced in Onsite?

**Comment by tywrenn** [comment]
After having a lot of experience with WordPress, when you create a plugin, you still have to maintain that plugin as methods can break when versions of WordPress change. Since no one has really tried to keep these plugins functional over the years, I see the best way forward in this is to drop WP portal support and instead create API functions that will allow clinics to create their own if they would like to.

**Comment by gutiersa** [comment]
I am sorry, I am late to this thread. I would like to volunteer to revive the WP portal and maintain it with help from, perhaps @CraigT543 
Here are some of my reasons:
1. Having options is a good idea, people/companies may provide their providers/doctors several options depending on their needs, or on how much they can spend.
2. I would like to have subscribers who are not necessarily my patients, who may eventually become my patients, hence I would like to continue to have that connectivity. Mainly, because I would like to educate my subscribers and if they decide to eventually request an appointment with me, they may. I am starting a practice and it will be a long time before I am able to accept insurance assignments. (the state is behind due to the pandemic) Some patients are waiting on me to take insurance as they cannot afford to pay themselves.
3. I would like to offer educational newsletters and maybe videos (most likely not) on healthcare issues. These will be available to my subscribers and patients. I would the ability to have my WP site to be integrated with openemr. I could maintain them for now.

If I don't do a good job, then you can delete it. 
Just send me the article and video tutorials such that I can figure this stuff out!

Thanks

**Comment by tywrenn** [comment]
@bradymiller Might wanna create a news post on this on the community page since this will effect a lot of people

**Comment by bradymiller** [comment]
I was under impression that this was non-functional and not updated. I can definitely add the wordpress portal back if it is actually functional and used. Or if somebody wishes to resuscitate it, can make a branch that has wordpress code in it on most recent codebase that devs can work on.

**Comment by gutiersa** [comment]
check out this thread. I think removing it will affect at least one person.

[https://community.open-emr.org/t/how-to-connect-wordpress-with-openemr/8937](url)

**Comment by bradymiller** [comment]
doh, sounds like I should add it back then. I'll make a PR to add it back and then see if that user is using it that way (will be a good way to see what changes that user may of made to get things to keep working).

**Comment by Bouguy** [comment]
Hi there, I would like to help to rebuild a link between OEMR and WP. May I help at least for test ? I know both solutions (as user and admin).

**Comment by CraigT543** [comment]
I have been using it for years.  I was a little late updating to 5.0.2.  I did that this week and now the WordPress portal does not work.  I get these errors:

`
2020-06-08T21:38:03-07:00 xxx [Mon Jun 08 21:38:03.496236 2020] [proxy_fcgi:error] [pid 21513:tid 140206211790592] [client xxx.xxx.xxx.x:xxx] AH01071: Got error 'PHP message: OpenEMR Error : Decryption failed authentication.', referer: https://xxx.xxx.xxx.x:xxx/openemr/interface/main/tabs/main.php?token_main=SKXnjgk40KbqabRYiRVFMB2HA5xs3bvaPHl3wvEL
2020-06-08T21:38:03-07:00 xxx [Mon Jun 08 21:38:03.496485 2020] [proxy_fcgi:error] [pid 21513:tid 140206211790592] [client xxx.xxx.xxx.x:xxxx] AH01071: Got error 'PHP message: PHP Warning:  Variable passed to each() is not an array or object in /path to/openemr/interface/cmsportal/list_requests.php on line 266PHP message: PHP Warning:  Variable passed to each() is not an array or object in /path to/openemr/interface/cmsportal/list_requests.php on line 267', referer: https://xxx.xxx.xxx.x:xxxx/openemr/interface/main/tabs/main.php?token_main=SKXnjgk40KbqabRYiRVFMB2HA5xs3bvaPHl3wvEL
`
In 5.0.1 I had similar problems and to work around I just used the older files from 5.0.0.  But it looks like in 5.0.2 those old files just will not work.  It looks like the issue is the authentication routine.  Something is definitely not working there.  I am not sure how to trace it down.  Very frustrating for me because now I have to enter patient information by hand again.  

I had made several modifications to the sunset patient portal in word press due to problems with Ninja.  My work around was to use Michael Simpson's CFDB (https://cfdbplugin.com/). So I rewrote my webserve to use Michael Simpson's interface which works with Ninja, CF7, and my favorite, Gravity Forms.  All worked well until this week.  So if someone can help me with the authentication issue I will freely share what I have working.

WordPress has a very nice API now and I think it would be wise to work with the API rather than using the Sunset Portal Method.  It works for now but I would not be surprised if it stops working in future iterations of WordPress.

**Comment by gutiersa** [comment]
It sounds like there may be encryption now in openemr, and that is why you may be getting the first error.
The other issue may be a PHP version issue. That is when I get those errors.

I definitely am interested in helping with the WP portal option. Mostly because I have experience with the portal from my previous job, I will learn the onsite portal, and then we can compare with the WP portal. We can bring all the nice features to both the onsite portal and the WP portal (yes, I agree with the API, but I would agree with us maintaining it for ease of download and install) such that users have options.

**Comment by CraigT543** [comment]
Ok. The encryption is not an issue.  The problem was that after I did the upgrade to 5.0.2 I needed to re-enter my credentials for the portal.  So now the authentication error is gone.  Understandable.  I should have thought to look at that before.

But I still am getting this:

`
2020-06-09T22:41:21-07:00 xxx [Tue Jun 09 22:41:21.656654 2020] [proxy_fcgi:error] [pid 21513:tid 140206036440832] [client xxx.xxx.xxx.x:xxxx] AH01071: Got error 'PHP message: PHP Warning:  Variable passed to each() is not an array or object in /my path to/openemr/interface/cmsportal/list_requests.php on line 266PHP message: PHP Warning:  Variable passed to each() is not an array or object in /my path to/openemr/interface/cmsportal/list_requests.php on line 267', referer: https://xxx.xxx.xxx.x:xxxx/openemr/interface/main/tabs/main.php?token_main=OxveEgIDZesA14Hsa9Fe1LCyjZoxfPOo6pcEBJoc
`

This was not a problem before with same code in 5.0.2  But now it is not recognizing the information received as an array.  I am looking at the Sunset Portal  code and I see arrays are built in foreach loops.  Looks like an arrray is being built.  And it worked in 5.0.1 so I am very confused at this response.

**Comment by mdsupport** [comment]
@tywrenn Since the original WP portal was contributed, the architecture of this project has evolved significantly. The environment in which the product operates has changed even more and will continue to do so. 
Addition of more presentation related components to the standard project makes it costlier to implement component version updates. When You and @sjpadgett went through BS4 upgrades, you end up treating each piece of code equally just because it is in the standard code base. I think core project should have more efforts directed towards **medical** functionality and making life of clinical users as easy as possible.
At the same time there is big need for each specialty to have a tailored outward looking presence that can be implemented by different skill sets. Since these should be loosely linked with core product I do not see much benefit by making it a "module".

**Comment by CraigT543** [comment]
When we moved to 5.0.1 the portal failed for me. I could see that encryption was changing.  But, I was able to just move over my old files using the older encryption system with crypto.php and it worked again.  I think /library/crypto.php was still running in 5.0.1 so I could just replace the new cmsportal files with the old ones from 5.0.0 and it continued to work.  Again in 5.0.2 the new cmsportal files did not sync with sunsetportal just like with 5.0.1.  So, I used the exact same files that were working for me from 5.0.0.  But because crypto.php was taken out for 5.0.2 and now cmsportal is a complete fail with the new encryption.  

I have been putting back some of the pieces of the old encryption system and little by little I am getting rid of errors related to credentials and file sync.  Now I have new errors indicating that I do not have all the configuration pieces in. For example the old decryption routine that was in c_document.class.php is now missing and I can add that.  We will see.  Could it be that the new encryption routine was not fully implemented in CMS portal for 5.0.1 or 5.0.2?  I can see that efforts were made to transition it in to the new system but was it tested? It never worked for me and it is not working for me now.  Because when I use the new system for some reason I get the "Variable passed to each() is not an array or object"  on lines 266 and 267 of list_requests.php.  That did not happen in 5.0.1 if I used my files from 5.0.0 and the only change between the two is removal of the routine for crypto.php.  It seems that the new encryption method was never fully worked out for the cmsportal because it was assumed to be dead (understandable).  But not dead to me.

**Comment by gutiersa** [comment]
Yes, I agree, It would be good if it remains included. My server is currently down. As soon as I fix it this weekend, I could help some.

**Comment by CraigT543** [comment]
I was not successful in getting crypto.php to function with 5.0.2.  The handshake is not taking place between the list_requests.php in openemr and webserve.php in the sunset portal.  The error log tells me I have a probable configuration error.  I am likely missing some dependency for crypto.php to work.  And I am also certain that the webserve application for the sunset portal is not set up to receive the csrf_token_form in the new method.  So the alternate option is to rewrite the webserve to accept the csrf token in the new routine.  I am not sure where to start there yet.

**Comment by bradymiller** [comment]
hi @CraigT543 ,
Just submitted a PR with all the wordpress portal code in it on OpenEMR most recent development codebase:
https://github.com/openemr/openemr/pull/3639

Are your encryption/decryption issues in the interface/cmsportal/upload_form.php script?

**Comment by CraigT543** [comment]
No the file that initiates the connection in OpenEMR  is  portal.inc.php with webserve.php in Wordpress.  As far as I can tell, portal.inc.php does a little cURL handshake with webserve.php and list_request.php grabs an array from webserve.php of all forms and messages that are on the wordpress site.  I am pretty sure it all is initiated with portal.inc.php.

**Comment by bradymiller** [comment]
Are you seeing errors in the handshake?
https://github.com/openemr/openemr/blob/daf9411666edafdfe3f3486c8fb3278dc9fd78f7/interface/cmsportal/portal.inc.php#L40
or
https://github.com/openemr/openemr/blob/daf9411666edafdfe3f3486c8fb3278dc9fd78f7/interface/cmsportal/portal.inc.php#L47

Note that the crypto call in that function is just decrypting the password that is stored in globals. So, the way to fix this would be to resave your wordpress password in globals, which will then get encrypted in database (and then correctly decrypted when collected from database).

**Comment by bradymiller** [comment]
Or is is erroring out in webserve.php at:
https://github.com/openemr/sunset-patient-portal/blob/master/webserve.php#L33
or
https://github.com/openemr/sunset-patient-portal/blob/master/webserve.php#L38
or
https://github.com/openemr/sunset-patient-portal/blob/master/webserve.php#L52

**Comment by CraigT543** [comment]
I was haveing some cURL errors after the update.  But I rebuilt my ssl certificate and that went away.  So I think that the handshake is working fine.  

But I am getting errors at the 'list'  and 'message' requests for the webserve.php in WordPress initiated by OpenEMR in list_requests.php and the error stops at line 155 in the 5.0.2 verson.  I think this is because of the added csrf_token_form variable.  If I use the 5.0.0 version of the files (with no csrf_token_form) it does not stop there but I get these errors:

PHP Warning:  Variable passed to each() is not an array or object in /web/openemr/interface/cmsportal/list_requests.php on line 268
PHP Warning:  Variable passed to each() is not an array or object in /web/openemr/interface/cmsportal/list_requests.php on line 269'

And those areas talk to the areas of the webserve.php that you have noted.

Craig

**Comment by CraigT543** [comment]
WHY HAVE A WORDPRESS PATIENT PORTAL
First, WordPress has so many plugins and themes that make website building rather quick and slick.  It is also rather easy to make a secure patient membership site for patient contact.
  
Some reasons I like the WordPress portal.  I have modified the Sunset Portal to sync with this plug in https://cfdbplugin.com/.  This allowed me to sync with GravityForms and CF7 as well as Ninja and many other Wordpress forms.  Ninja is good but it is not the best.  Ninja has a free option but CF7 is free and will do more.  GravityForms will do more than Ninja for less money.  That being said, with the wordpress forms tools I have the ability to do complex on line assessments with calculations and weighted scales.  It is easy to build them.  For lengthy assessments clients can save and return to the form in GravityForms.  Then the pdf version of the form can sync into OpenEMR and save with the client.  That is fantastic.  This is what I use the portal for.    

I was syncing messages with Cartpauj PM but it stopped working.  I then updated the webserve to work with Frontend PM which started off as a improvement to Cartpauj but is now something much more complex.  Frontend PM updated about a year ago and my sync stopped on openEMR I have not bothered to fix it.  The syncing of messages was not too good with the patient portal.  The messages would sync in but you had to save them by cut and paste.  So it was not finished.  We need a better method for syncing Private Messages.
 
Wordpress has several e-commerce solutions that work great for payments.  Woocommerce integrates with almost all the major systems.  And you can set it up so that, paypal for example, only contains cryptic information about the appointment on their website.  When a person books with me for example paypal just shows my website address and the amount they paid.  But if they log into my site as a user they can get a detailed invoice of all their appointments.  I still do superbills through OpenEMR but very fue of my clients are seeking reimbursement.  With wordpress I am not locked in to any one payment method.  I can shop around for best rates and switch easily.  We cannot with OpenEMR.

There are multiple appointment options with Wordpress.  I have used easy appointments for years.  It has a nice API.  I like it a lot but it is hard to integrate that with paypal.  I have switched over to Amilia Appointments and am very happy with it.  It integrates completely with Woocommerce.  The recurring appointments feature is not so good but other than that, with Amelia, my clients can book appointments, pay at the time of booking the appointment, cancel, and reschedule.  And with the Woo-commerce dashboard they are able to pay invoices for appointments that I have booked on my end.  I can set up different user roles so that when a client logs in who is an insurance patient (very few of those) I can have a special rate for them rather than others.  I have set up about 4 different user roles.  One allows certain clients options that are not available to others etc.  All this is easy to do with WordPress.  And it is all done with one log in.

Unfortunately, the appointments and the private messaging do not sync with OpenEMR.  And I have to manually archive and delete messages and appointments that are older than 6 months to avoid having too much patient data available on my website.  When the Sunset Patient portal was running, it would delete forms from my website as soon as I loaded them into OpenEMR.  I really liked that.  So there needs to be a better option. 

But also I have integrated GoToMeeting with my site.  Zoom integrates very easily with Amilia Appointments.  And these components update easily.

I believe that more people would use OpenEMR if we had a plugin that was available on the https://wordpress.org/plugins/ site.  The Sunset Patient Portal was not available on in the WordPress Plugins site.  I think Cartpauj PM was but it no longer is because it stopped working with WordPress.  So that is the key, getting the plugin into the https://wordpress.org/plugins/ site.  That would lead more people to OpenEMR as an option.  

For mental health, there are lots of one size fits all EMR solutions that are really confining.  I have tried them and find the best way to go for me has been OpenEMR with a WordPress site.  This gives me a full service EMR with a very complex and useful patient portal.

**Comment by bradymiller** [comment]
Goals are to definitely get it back in 6.0.0 and working :)

The csrf_token_form shouldn't be an issue. When you mean it stops at line 155 in the 5.0.2 version, what were the specific errors reported?

**Comment by jfischburg-lifemesh** [comment]
I would like to check-in on the progress of this initiative and contribute. @bradymiller @CraigT543 @gutiersa @tywrenn @sjpadgett @sunsetsystems what is the current status, and what needs doing?

**Comment by Paul-B** [comment]
Hey all,

I've been reading this thread and all the others I could find regarding the connection between openemr and wp. At the moment I run a web-facing client registration site on wp that uses formidableforms to register users and collect data. The basic user data is saved in de wordpress user table along with some meta data like a billing address and a client number. The sensitive stuff is transferred to my private server wia the formidable wordpress api.
On the private server I run a simple client database that I made in wordpress. Now this is where openemr appears on the horizon. I am at a tilting point to decide to add functions that are already present in openemr like encounters and nice summaries that can be edited on the fly. Now if only openemr was able to sync to the wordpress user database then this would allow me to combine it all and do the real client care in openemr. The benefit of this is that there are obviously more people developing openemr. I see talk of api's in openemr and I think that is the way but it looks like I came here too early as it doesn't seem to be present.

**Comment by CraigT543** [comment]
Hello Bradly,

I just noticed you asked for error messages.  Sorry for the delay.  This is all that I get:

`
Got error 'PHP message: PHP Warning:  Variable passed to each() is not an array or object in /pathtoopenemr/interface/cmsportal/list_requests.php on line 266PHP message: PHP Warning:  Variable passed to each() is not an array or object in /pathtoopenemr/interface/cmsportal/list_requests.php on line 267', referer: https://000.000.000.0:000/openemr/interface/main/tabs/main.php?token_main=MyTokenHere
`
jfischburg-lifemesh: As far as I know the current status is what you see here.  There is a problem with the token.

Paul-B: The webserve application on the word press side was only designed to work with Ninja forms and it is out dated.  The problem with Ninja is that they constantly change their methods.  I reworked the webserve to talk with the CFDB plugin found here: https://github.com/mdsimpson/contact-form-7-to-database-extension/releases.  This plug in integrates with most of the major word press forms plugins including Formidable, Ninja, Gravity, CF7 and many others.  I like the flexibility but I am not sure if this is the best approach because I am depending on CFDB to continue to be developped and the developer almost gave it up a couple years ago.  It would be nice if we had a web server that was as flexible to replace the sunset systems one we have now.  But for now (or at least until 5.0.1) my modification worked fine with OpenEMR to do the sync.  And the sync is a little problematic.  It is slow.  And the messaging application no longer works so we need another solution there.  I have been using FrontendPM Pro which really has been very good for private messaging.  I had it syncing for a while with the webserve but then FrontendPM Pro made some improvements that broke my code and I have not had time to fix that yet. Until we have a solution for the token, I cannot work on that.

**Comment by CraigT543** [comment]
Now that I am getting around to looking I see that line 267 is the sync for messages and the messages portion is problematic for various reasons.  The first issue is that there is no Cartpauj PM plugin anymore.  And my sync with Frontend PM also stopped working so I will dig in to this.

**Comment by alsharabi1962** [comment]
> The first issue is that there is no Cartpauj PM plugin

could the plug in installed manually on WP after downloading the files from:
https://github.com/sunsetsystems/cartpauj-pm

**Comment by adunsulag** [comment]
Is there still a desire for individuals to implement this?  The issue is three years old and we haven't seen any action on it. I would recommend we close this issue and if anyone wants to re-open it in the future they can.

**Comment by gutiersa** [comment]
Sure, I agree. I never was able to implement WP. The OpenEMR portal is very good.

**Comment by sjpadgett** [comment]
awww shucks(boy digging toe in to sand) Sandra!

**Comment by CraigT543** [comment]
I think I may be one of the only ones who successfully had it implemented and loved it.  I have never been able to transition to the OpenEMR portal because my clients are so used to my current secure message service (FrontendPM) and forms applications (Gravity Forms), and scheduling software (Amelia) in WordPress.  These applications are all so customizable and flexible. I just wish they talked to OpenEMR.  I have wanted to build an interface.  They all have nice REST APIs.  But for now I am just using them side by side.  I will get around to it some day.  I had it working with FrontendPM and Gravity Forms for a while but then Sunset Portal was just so out of date was impossible.  Disappointing.

**Comment by kojiromike** [comment]
The WordPress patient portal was consolidated/deprecated in #3593 (merged June 2020). Closing as the standalone WordPress portal is no longer maintained. If there are remaining portal issues, please open a new issue against the current patient portal.
